unit_tests/ngp_algorithms/UnitTestNodalGradPOpenBoundary.C

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019 National Technology & Engineering Solutions of Sandia, LLC
+// (NTESS), National Renewable Energy Laboratory, University of Texas Austin,
+// Northwest Research Associates. Under the terms of Contract DE-NA0003525
+// with NTESS, the U.S. Government retains certain rights in this software.
+//
+// This software is released under the BSD 3-clause license. See LICENSE file
+// for more details.
+//
+
+#include "kernels/UnitTestKernelUtils.h"
+
+#include "UnitTestHelperObjects.h"
+
+#include "ngp_algorithms/UnitTestNgpAlgUtils.h"
+#include "ngp_algorithms/NodalGradPOpenBoundaryAlg.h"
+#include "ngp_algorithms/GeometryBoundaryAlg.h"
+#include "ngp_algorithms/NodalGradAlgDriver.h"
+#include "ngp_algorithms/GeometryAlgDriver.h"
+
+TEST_F(LowMachKernelHex8Mesh, NGP_nodal_grad_popen)
+{
+  // Only execute for 1 processor runs
+  if (bulk_.parallel_size() > 1) return;
+
+  const std::string meshSpec = "generated:20x20x20";
+  const bool doPerturb = false;
+  const bool generateSidesets = true;
+
+  fill_mesh_and_init_fields(meshSpec, doPerturb, generateSidesets);
+
+  unit_test_utils::HelperObjects helperObjs(bulk_, stk::topology::HEX_8, 1, partVec_[0]);
+
+  const double zero= 0.0;
+  const double one = 1.0;
+  const double two = 2.0;
+  const double oneVecTwelve[12] = {one, one, one, one, one, one, one, one, one, one, one, one};
+  stk::mesh::field_fill(zero, *dpdx_);
+  stk::mesh::field_fill(two, *dnvField_);
+  stk::mesh::field_fill(one, *pressure_);
+  stk::mesh::field_fill_component(oneVecTwelve, *exposedAreaVec_);
+
+  stk::mesh::Part* surface1 = meta_.get_part("surface_1");
+  {
+    sierra::nalu::GeometryAlgDriver geomAlgDriver(helperObjs.realm);
+    geomAlgDriver.register_face_algorithm<sierra::nalu::GeometryBoundaryAlg>(
+      sierra::nalu::WALL, surface1, "geometry");
+    geomAlgDriver.execute();
+  }
+
+  stk::mesh::field_fill(2.0, *dnvField_);
+
+  {
+    helperObjs.realm.solutionOptions_->activateOpenMdotCorrection_ = true;
+    sierra::nalu::ScalarNodalGradAlgDriver algDriver(helperObjs.realm, "dpdx");
+    algDriver.register_face_elem_algorithm<sierra::nalu::NodalGradPOpenBoundary>(
+      sierra::nalu::OPEN, surface1, stk::topology::HEX_8, "nodal_grad_pressure_open_boundary", false);
+    algDriver.execute();
+  }
+
+  {
+    const double dpdxref[6][3] = {{-0.25, 0.0,  -0.25},
+                                  {-0.25, 0.0,   0.25},
+                                  {-0.25,-0.25,  0.0},
+                                  {-0.25, 0.25,  0.0},
+                                  {-0.5,  0.0,   0.0},
+                                  { 0.0,  0.0,   0.0}};
+
+    const double tol = 1.0e-16;
+
+    ngp::Field<double> ngpdpdx(bulk_, *dpdx_);
+    ngp::Field<double> ngpcord(bulk_, *coordinates_);
+    ngpdpdx.modify_on_device();
+    ngpdpdx.sync_to_host();
+    ngpcord.modify_on_device();
+    ngpcord.sync_to_host();
```

Comment:
I don't understand why this is necessary.

Reply:
Shreyas,
Yes, syncing the coordinates back to the host is unnecessary.  I was just syncing all of the fields and included the coordinates even though they started on the host to begin with.
(I wish the email version of these comments included the previous comments)